content/js1/sprints/3/prep/index.md

```diff
@@ -8,4 +8,634 @@ backlog= 'Module-JS1'
+So we must use another method to extract _multiple_ characters from the given string.
+
+### ✂️ Extracting a slice
+
+We can use a function called `slice` to extract 1 or more characters from a string.
+
+[`slice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/slice) is a function that can take 2 arguments: a start index and an end index. `slice` will return a section of the string from the start index up to but not including the start index.
+
+```js
+time; // holds the value "23:00"
+```
+
+```js
+time.slice(0, 2); // will access the characters below
+```
+
+|   index   |   0   |   1   |
+| :-------: | :---: | :---: |
+| character | `"2"` | `"3"` |
+
+So `time.slice(0,2)` will evaluate to `"23"` when the `time` is `"23:00"`. Finally we must convert `"23"` to the number `23`, otherwise we can't compare this value properly. We can use the `+` operator to convert strings to numbers by writing the following:
+
+```js
++time.slice(0, 2); // evaluates to 23
+```
+
+### 🏗️ Assembling the parts
+
+Earlier we had an `if` statement in the form:
+
+```js
+if (expressionForHours > 12) {
+}
+```
+
+So now we've found an expression for the `hours`, we can write:
+
+```js
+if (+time.slice(0, 2) > 12) {
+}
+```
+
+If the time is `"23:00"` then the condition `+time.slice(0,2) > 12` will evaluate to `true` and the code inside the curly braces will be executed.
+
+This `if` statement is implementing the following part of the diagram from earlier:
+
+```mermaid
+
+flowchart TD
+
+A{Check: Are the hours greater than 12?}
+
+```
+
+Now we can format the string using our approach from earlier:
+we’ll need to append `"pm"` to the string expression and subtract 12 from the hours. So we get the following:
+
+```js
+if (+time.slice(0, 2) > 12) {
+  return `${time.slice(0, 2) - 12}:00 pm`;
+}
+```
+
+The `return` statement above implements the following steps we set out earlier:
+
+```mermaid
+
+flowchart TD
+
+D[Step 4: subtract 12 from the hours]
+D --> E[Step 5: add 'pm' to the rest of the time]
+E --> F[Step 6: return the new time]
+
+```
+
+Now we can **re-run** our assertions from earlier to check our function behaves as target.
+
+### 🧹 Improving the code
+
+Now the assertions pass: in other words, our function’s current output matches with the target output described in the assertions.
+
+However, at the moment, we’re making use of the same expression twice: `+time.slice(0,2)`. This means we’re calling the function `slice` twice. Additionally, expressions embedded inside curly braces and parentheses can often be difficult to read. In this situation it makes sense to label the recurring expression so we can reuse wherever it we need to in our code.
+
+Let’s create a variable called `hours` and assign to it our expression.
+
+```js
+function formatAs12HourClock(time) {
+  const hours = +time.slice(0, 2);
+
+  if (hours > 12) {
+    return `${hours - 12}:00 pm`;
```

Comment:
We should probably also also have a section where we talk about how this code is buggy because it assumes there are no minutes, and prompt the trainees to add a new assertion and fix that themselves. They already have all of the building-blocks they need to do this, so it should just be a matter of re-applying those same concepts around slicing.

I'd be tempted to also set a piece of coursework around considering the classes of input, and interesting edge cases, they would maybe want to test. Let's note in that exercise also that there are two major ways of identifying these classes/edge-cases: "Domain-driven" (i.e. "morning vs afternoon", "midnight and midday", etc), and "Implementation-driven" (e.g. remembering that we need to include the correct minutes isn't necessarily a core _domain_ problem, but we _can_ identify it as an interesting case to test based on knowing what our implementation looks like - we're separating the hours from minutes, so may need to pay extra attention to the idea that we may do something wrong with the minutes). This would be a great discussion for TAs to facilitate on the Saturday.